common/src/main/java/com/google/auto/common/MoreTypes.java

```diff
@@ -139,24 +139,32 @@ public String toString() {
       this.bArguments = bArguments;
     }
 
+    protected boolean canEqual(@Nullable Object o) {
+      return o instanceof ComparedElements;
+    }
+
     @Override
     public boolean equals(@Nullable Object o) {
-      if (o instanceof ComparedElements) {
-        ComparedElements that = (ComparedElements) o;
-        int nArguments = aArguments.size();
-        if (!this.a.equals(that.a) || !this.b.equals(that.b) || nArguments != bArguments.size()) {
-          // The arguments must be the same size, but we check anyway.
-          return false;
-        }
-        for (int i = 0; i < nArguments; i++) {
-          if (aArguments.get(i) != bArguments.get(i)) {
-            return false;
-          }
-        }
-        return true;
-      } else {
+      if ( !(o instanceof ComparedElements) ) return false;
+      ComparedElements that = (ComparedElements) o;
+      if ( !that.canEqual(this) ) return false;
+
+      int nArguments = this.aArguments.size();
+      if ( nArguments != that.aArguments.size() ) return false;
+      // The arguments must be the same size, but we check anyway.
+      if ( nArguments != this.bArguments.size() || nArguments != that.bArguments.size() ) return false;
+
+      if ( !this.a.equals(that.a) || !this.b.equals(that.b) )
         return false;
+
+      for (int i = 0; i < nArguments; i++) {
+        if ( !this.aArguments.get(i).toString().equals( that.aArguments.get(i).toString() ) )
```

Comment:
Just to clarify, should I do something like this
```java
this.aArguments.get(i) != that.aArguments.get(i) )
```
instead?

 If so, is it ok (and do you prefer) if I put your explanation as a comment in the code?

Reply:
Yes, that's what I meant, and yes, I think it would make sense to put at least the two sentences starting with "The purpose here" as a comment. This is quite tricky code and should have more comments than it does.